Support for KNP when using `TranslationFieldFilter`

Liinkiing opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 2.x is for everything backwards compatible, like patches, features and deprecation notices
    - master is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataTranslationBundle/blob/2.x/CONTRIBUTING.md#the-base-branch
-->
I am targeting this branch, because the `TranslationFieldFilter` was not working when using **KnpLabs**. The query made in the filter was not handling the database schema of **KnpLabs**. This PR adds support for both of these providers.

It adds a global option for the bundle `default_filter_mode`, which can either be `gedmo` or `knplabs`. And it adds a `filter_mode` options to the field filter, to allow customizing it on a per-field basis.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

I could not find a related issue

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataTranslationBundle/releases,
    please keep it short and clear and to the point
-->

<!-- 
    If you are updating something that doesn't require
    a release, you can delete the whole Changelog section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Added a global option `default_filter_mode` to change every filter mode
- Added `filter_mode` option for `TranslationFieldFilter` to change the filtering mode based on the provider (either `knplabs` or `gedmo`)
```

<!--
    If this is a work in progress, uncomment this section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
    
    ## To do
    
    - [ ] Update the tests
    - [ ] Update the documentation
    - [ ] Add an upgrade note
-->

core23 commented on src/Filter/TranslationFieldFilter.php:
```diff
@@ -16,11 +16,19 @@
 
 final class TranslationFieldFilter extends Filter
 {
+    private $filterMode;
```
Please add phpdoc

core23 commented on src/Filter/TranslationFieldFilter.php:
```diff
@@ -16,11 +16,19 @@
 {
+    private $filterMode;
+
+    public function __construct(string $filterMode)
```
Can you please add a default value for this

core23 commented on src/Filter/TranslationFieldFilter.php:
```diff
@@ -54,22 +61,40 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
-        $this->applyWhere($queryBuilder, $queryBuilder->expr()->orX(
-            $queryBuilder->expr()->andX(
-                $queryBuilder->expr()->eq($joinAlias.'.field', $queryBuilder->expr()->literal($field)),
+        if (TranslationFilterMode::GEDMO === $filterMode) {
```
This method is very long, please extract some methods

core23 replied:
Ping @Liinkiing

core23 commented on tests/Traits/GedmoOrmTest.php:
```diff
@@ -88,7 +88,7 @@ public function testTranslationFieldFilter()
         $builder = new ProxyQuery($qb);
 
-        $filter = new TranslationFieldFilter();
+        $filter = new TranslationFieldFilter('gedmo');
```
No asserts for the new gedmo type?

Liinkiing replied:
What did you mean exactly?

Liinkiing commented:
Yes sorry, first time contributing 😳, It's still a WIP but I miss clicked and did not set it as a draft 🤐

core23 commented:
> Yes sorry, first time contributing 😳

It's okay. Everybody was new at a time 👍 It's just a feedback. No need to hurry

Liinkiing commented:
@core23 I have a small question, is there any reason why in `GedmoOrmTest`, the const `ARTICLE` and `TRANSLATION` have their FQCN written in plain text and not using `::class` notation? Is it for compatibility? Do I have to do the same in mode test case? 

By the way I've updated my PR based on your review and added tests 👍

core23 commented:
> @core23 I have a small question, is there any reason why in `GedmoOrmTest`, the const `ARTICLE` and `TRANSLATION` have their FQCN written in plain text and not using `::class` notation? Is it for compatibility? Do I have to do the same in my test case?

Technical dept :laughing: This was not possible in PHP < 7, but you can use it since a while

Liinkiing commented:
@core23 any feedbacks on the updated code? I've made the changes you requested last time 👍

core23 commented on tests/Fixtures/Traits/ORM/Knplabs/ArticleTranslation.php:
```diff
@@ -0,0 +1,55 @@
+ * @ORM\Table(name="article_translation")
+ * @ORM\Entity
+ */
+class ArticleTranslation
```
Please make all new test classes final. After that, it is RTM

Sorry for the delay @Liinkiing

Liinkiing replied:
No problem 👍, I'll do it, do you want me to rebase my branch ?

greg0ire commented:
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/2.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

Liinkiing commented:
@greg0ire Done! 👍

greg0ire commented on tests/Traits/KnplabsOrmTest.php:
```diff
@@ -0,0 +1,117 @@
+    /** @var TranslatableListener */
+    private $translatableListener;
+
+    protected function setUp()
```
The tests will fail on Travis unless you provide a return type declaration.
```suggestion
    protected function setUp(): void
```

franmomu commented on src/DependencyInjection/Configuration.php:
```diff
@@ -46,6 +49,16 @@ public function getConfigTreeBuilder()
                     ->defaultValue('en')
                     ->info('The frontend locale that is used by default.')
                 ->end()
+                ->scalarNode('default_filter_mode')
```
What about using [enumNode](https://symfony.com/doc/4.4/components/config/definition.html#enum-nodes)? It would simplify this a little bit.

Liinkiing replied:
Didn't know about this one! Thanks, will be using this

Liinkiing commented:
@greg0ire @franmomu I've made the changes you requested 👍

greg0ire commented on docs/reference/orm.rst:
```diff
@@ -178,7 +178,10 @@ Configure search filter
-filter to search on fields and on their translations.
+filter to search on fields and on their translations. Depending on whether you choose to use **KnpLabs** or **Gedmo**,
+you should configure the ``default_filter_mode`` in the configuration. You can also configure how
+the filtering logic should works on a per-field basis by specifying an option named ``filter_mode`` on your field.
```
```suggestion
the filtering logic should work on a per-field basis by specifying an option named ``filter_mode`` on your field.
```

greg0ire commented on src/DependencyInjection/Configuration.php:
```diff
@@ -21,6 +22,8 @@
  */
 class Configuration implements ConfigurationInterface
 {
+    private const AVAILABLE_FILTER_TYPES = [TranslationFilterMode::GEDMO, TranslationFilterMode::KNPLABS];
```
Wouldn't that constant be better in `TranslationFilterMode`?

OskarStark replied:
And please linebreak the array

OskarStark commented on src/DependencyInjection/Configuration.php:
```diff
@@ -46,6 +49,11 @@ public function getConfigTreeBuilder()
                     ->info('The frontend locale that is used by default.')
                 ->end()
+                ->enumNode('default_filter_mode')
+                    ->values(self::AVAILABLE_FILTER_TYPES)
```
Hmm we should use the constants here and remove the private const. it’s only used once

OskarStark commented on src/Filter/TranslationFieldFilter.php:
```diff
@@ -35,16 +46,15 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
             foreach ($joins as $join) {
                 if ($join->getAlias() === $joinAlias) {
                     $aliasAlreadyExists = true;
-
```
Please keep this line

OskarStark commented on src/Filter/TranslationFieldFilter.php:
```diff
@@ -54,22 +64,21 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
-        $this->active = true;
+            $this->active = true;
+        } else {
+            throw new \LogicException(
```
Please keep it in one line it’s not to long

OskarStark commented on tests/Fixtures/Traits/ORM/Knplabs/Article.php:
```diff
@@ -0,0 +1,51 @@
+        return $this->id;
+    }
+
+
```
Please remove these 2 blank lines

OskarStark commented on tests/Traits/KnplabsOrmTest.php:
```diff
@@ -0,0 +1,117 @@
+    {
+        parent::setUp();
+
+        if (!class_exists('Doctrine\\ORM\\Version')) {
```
Can you use ::class constant here?

OskarStark commented on tests/Traits/KnplabsOrmTest.php:
```diff
@@ -0,0 +1,117 @@
+ */
+class KnplabsOrmTest extends DoctrineOrmTestCase
+{
+    public const ARTICLE = Article::class;
```
Private?

OskarStark commented on tests/Traits/KnplabsOrmTest.php:
```diff
@@ -0,0 +1,117 @@
+class KnplabsOrmTest extends DoctrineOrmTestCase
+{
+    public const ARTICLE = Article::class;
+    public const TRANSLATION = ArticleTranslation::class;
```
Private?

OskarStark commented on tests/Fixtures/Traits/ORM/Knplabs/ArticleTranslation.php:
```diff
@@ -0,0 +1,55 @@
+    {
+        return Article::class;
+    }
+
```
Please remove blank line

Liinkiing commented:
@OskarStark @greg0ire Made the new changes! 👍 (don't know if I should ping you everytime, I don't want to spam you 👼)

OskarStark commented:
No problem ✌🏻🤗

franmomu commented on tests/Model/KnplabsTest.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @author Omar Jbara <omar.jbara2@gmail.com>
+ */
+class KnplabsTest extends TestCase
```
```suggestion
final class KnplabsTest extends TestCase
```

franmomu commented on tests/Traits/KnplabsOrmTest.php:
```diff
@@ -0,0 +1,118 @@
+/**
+ * @author Omar Jbara <omar.jbara2@gmail.com>
+ */
+class KnplabsOrmTest extends DoctrineOrmTestCase
```
```suggestion
final class KnplabsOrmTest extends DoctrineOrmTestCase
```

greg0ire commented on tests/Model/KnplabsTest.php:
```diff
@@ -0,0 +1,33 @@
+ */
+final class KnplabsTest extends TestCase
+{
+    public function testTranslatableModel()
```
```suggestion
    public function testTranslatableModel(): void
```

greg0ire commented:
> don't know if I should ping you everytime

Makes no difference to me, I get a notification either way :P

Liinkiing commented:
@core23 hey, is it good for you? I've made all the requested changes 👍

greg0ire commented:
Thanks @Liinkiing
